# Introduction

In this notebook, I evaluate a base (non–fine-tuned) code-fixing model on my dataset to establish a baseline. For each sample, the model receives the task description and the buggy Python code and returns a proposed fix. I then run the generated code in an isolated subprocess and record whether it executes successfully within a time limit. If the first attempt fails (except timeouts), I make a second attempt by providing the runtime error back to the model.

At the end, I compute and print summary metrics (e.g., runtime success rate, attempt usage, and model latency). The in-memory results table can be inspected directly in the notebook and will later be compared with the fine-tuned model using the same evaluation split.

## 1. Setup

In this section, I import the required libraries and configure the runtime.  
This includes standard utilities, dataset handling, loading environment variables (for the OpenRouter API key), a notebook-friendly progress bar, and an SSL trust-store fix on Windows to ensure OpenRouter requests work reliably. I also import the LangChain components used to call the model with system/user messages.


In [14]:
# -----------------------
# Standard library
# -----------------------
import os
import json
import re
import time
import random
import subprocess
import sys
import tempfile
from pathlib import Path

# -----------------------
# Data / datasets
# -----------------------
import pandas as pd
from datasets import Dataset

# -----------------------
# Environment variables
# -----------------------
from dotenv import load_dotenv

# -----------------------
# Progress bar (notebook-friendly)
# -----------------------
from tqdm.auto import tqdm


In [ ]:
# -----------------------
# SSL / certificates 
# -----------------------
import truststore
truststore.inject_into_ssl()

In [ ]:
# -----------------------
# Networking (optional: for catching request-related errors)
# -----------------------
from requests.exceptions import RequestException

# -----------------------
# LLM client (OpenRouter via LangChain)
# -----------------------
from langchain_openai import ChatOpenAI
from langchain_core.messages import SystemMessage, HumanMessage

## 2. Load and split the dataset

Here I load the JSON dataset from the local project directory, set a fixed random seed for reproducibility, and convert the list of samples into a Hugging Face `Dataset`. I then create a train/eval split (15% for evaluation) that will be reused later to compare baseline and fine-tuned performance on the same eval set.


In [4]:
SEED = 42
random.seed(SEED)

PROJECT_ROOT = Path.cwd()  
DATA_PATH = (PROJECT_ROOT.parent / "Datasets" / "final_dataset.json").resolve()

with open(DATA_PATH, "r", encoding="utf-8") as f:
    data = json.load(f)


print("Samples:", len(data))
print("Keys:", list(data[0].keys()))

dataset = Dataset.from_list(data)
split = dataset.train_test_split(test_size=0.15, seed=SEED)
train_dataset = split["train"]
eval_dataset  = split["test"]

print("Train:", len(train_dataset))
print("Eval :", len(eval_dataset))

Samples: 582
Keys: ['title', 'description', 'difficulty', 'correct_code', 'incorrect_code', 'error_type']
Train: 494
Eval : 88


## 3. Configure the model client

In this cell, I load the OpenRouter API key from a local `.env` file (so it is not stored in the notebook), configure the OpenRouter endpoint and model ID, and initialize a LangChain `ChatOpenAI` client. I also define `llm_fix()`, a small wrapper that measures request latency and retries with exponential backoff to handle transient connection or rate-limit issues.


In [5]:
load_dotenv()  # loads variables from .env into environment

OPENROUTER_API_KEY = os.getenv("OPENROUTER_API_KEY")
if not OPENROUTER_API_KEY:
    raise ValueError("OPENROUTER_API_KEY not found. Check your .env file.")

OPENROUTER_BASE_URL = os.getenv("OPENROUTER_BASE_URL", "https://openrouter.ai/api/v1")
MODEL_ID = os.getenv("OPENROUTER_MODEL", "qwen/qwen2.5-coder-7b-instruct")

APP_REFERER = "http://localhost"
APP_TITLE = "code-fixer-eval"

llm = ChatOpenAI(
    model=MODEL_ID,
    api_key=OPENROUTER_API_KEY,
    base_url=OPENROUTER_BASE_URL,
    temperature=0.0,
    max_tokens=7000,
    default_headers={
        "HTTP-Referer": APP_REFERER,
        "X-Title": APP_TITLE,
    },
)

def llm_fix(messages, retries=6, min_backoff=1.0, max_backoff=20.0):
    """
    Calls llm.invoke(messages) with retry/backoff for connection/rate/transient errors.
    Returns (content, latency_seconds).
    """
    t0 = time.perf_counter()
    last_err = None

    for attempt in range(retries):
        try:
            resp = llm.invoke(messages)
            t1 = time.perf_counter()
            return resp.content, (t1 - t0)

        except Exception as e:
            # Catch common transient errors broadly (APIConnectionError, Timeout, 429, 5xx)
            last_err = e

            # backoff + jitter
            sleep_s = min(max_backoff, min_backoff * (2 ** attempt)) + random.random()
            print(f"[WARN] LLM call failed ({type(e).__name__}): {e} | retrying in {sleep_s:.1f}s...")
            time.sleep(sleep_s)

    raise last_err

## 4. Prompt templates

This section defines the system instruction and the user prompt format used to ask the model to fix buggy AI/ML Python code. I run two attempts per sample: the first uses only the task + buggy code, and the second (if needed) includes the runtime error message to guide a minimal correction.


In [6]:
# -----------------------
# Prompts
# -----------------------
SYSTEM_TEXT = (
    "You are a Python code fixer for AI/ML projects. "
    "Fix the provided code so it runs end-to-end and matches the task requirements. "
    "Rules: make the smallest possible change(s); do NOT add new features, demo blocks, or extra prints; "
    "do NOT refactor or rename unless required. "
    "Output ONLY the corrected Python code (no markdown, no explanations)."
)

def build_user_prompt_no_hint(ex):
    return (
        f"Task title: {ex['title']}\n"
        f"Task description: {ex['description']}\n\n"
        "Buggy code:\n"
        "```python\n"
        f"{ex['incorrect_code']}\n"
        "```\n\n"
        "Fix the code. Output ONLY the corrected Python code."
    )

def build_user_prompt_with_error(ex, prev_code, error_text):
    return (
        f"Task title: {ex['title']}\n"
        f"Task description: {ex['description']}\n\n"
        "The following code still fails at runtime with this error:\n"
        f"{error_text}\n\n"
        "Code:\n"
        "```python\n"
        f"{prev_code}\n"
        "```\n\n"
        "Fix ONLY what is necessary to remove the error and satisfy the task. "
        "Do not add new sections. Output ONLY the corrected Python code."
    )


## 5. Helper utilities

This section contains utility functions used during evaluation. It includes a robust extractor to ensure the model output is treated as pure Python code, a sandboxed subprocess runner with timeouts and safe decoding, parsers for the dataset’s expected error metadata, and small helpers for classifying runtime failures and selecting per-sample time limits.


In [ ]:
# -----------------------
# Helpers
# -----------------------

_CODEBLOCK_RE = re.compile(r"```(?:python)?\s*(.*?)```", re.DOTALL | re.IGNORECASE)

def extract_python_code(text: str) -> str:
    """
    Best-effort extraction of pure Python code from an LLM response.
    Priority:
      1) content inside ```python ... ```
      2) otherwise, remove obvious non-code leading lines until we hit code-like content
    """
    text = (text or "").strip()

    # 1) Prefer fenced code if present
    m = _CODEBLOCK_RE.search(text)
    if m:
        return m.group(1).strip()

    # 2) Remove leading junk lines until it looks like code
    lines = text.splitlines()

    def looks_like_code(line: str) -> bool:
        s = line.strip()
        if not s:
            return False
        # common python starters
        if s.startswith(("import ", "from ", "def ", "class ", "@", "if ", "for ", "while ", "try:", "with ", "print(")):
            return True
        # assignments / function calls
        if re.match(r"^[A-Za-z_][A-Za-z0-9_]*\s*=", s):
            return True
        if re.match(r"^[A-Za-z_][A-Za-z0-9_]*\(", s):
            return True
        # comments can be part of code
        if s.startswith("#"):
            return True
        return False

    start = 0
    for i, line in enumerate(lines):
        if looks_like_code(line):
            start = i
            break
    cleaned = "\n".join(lines[start:]).strip()

    # 3) If model echoed "python" label, remove it
    cleaned = re.sub(r"^\s*python\s*\n", "", cleaned, flags=re.IGNORECASE)

    return cleaned


def run_python(code: str, timeout=25, debug=False):
    code = extract_python_code(code)

    with tempfile.TemporaryDirectory() as td:
        tmp = Path(td) / "_tmp_run_eval.py"
        tmp.write_text(code, encoding="utf-8")

        cmd = [sys.executable, str(tmp)]
        if debug:
            print("CMD:", cmd)
            print("TMP exists:", tmp.exists(), "size:", tmp.stat().st_size)

        try:
            proc = subprocess.run(
                cmd,
                stdout=subprocess.PIPE,        # capture stdout for debugging
                stderr=subprocess.PIPE,
                text=True,
                encoding="utf-8",
                errors="replace",
                timeout=timeout,
            )
            ok = (proc.returncode == 0)
            err = (proc.stderr or "")
            err_tail = "\n".join(err.splitlines()[-25:])
            if debug and not ok:
                print("STDOUT tail:\n", "\n".join((proc.stdout or "").splitlines()[-25:]))
                print("STDERR tail:\n", err_tail)
            return ok, err_tail

        except subprocess.TimeoutExpired:
            return False, f"TimeoutExpired: exceeded {timeout}s"
        except Exception as e:
            # show the real reason
            return False, f"RunnerError: {type(e).__name__}: {e}"
        
        
def parse_error_type_field(error_type_text: str):
    """
    Field example:
      "NameError: ... || line: y_pred = best_modell.predict(X_test)"
    Returns:
      expected_exception="NameError" (or None)
      expected_line="y_pred = best_modell.predict(X_test)" (or None)
    """
    if not error_type_text:
        return None, None

    expected_exception = None
    expected_line = None

    m = re.match(r"\s*([A-Za-z_][A-Za-z0-9_]*)\s*:", error_type_text)
    if m:
        expected_exception = m.group(1)

    m2 = re.search(r"\|\|\s*line:\s*(.*)\s*$", error_type_text)
    if m2:
        expected_line = m2.group(1).strip()

    return expected_exception, expected_line


def contains_expected_bug_line(code: str, expected_line: str):
    if not expected_line:
        return None
    code = extract_python_code(code)
    return expected_line.strip() in code


def timeout_for(difficulty: str) -> int:
    d = (difficulty or "").lower()
    if d == "easy":
        return 120
    if d == "medium":
        return 180
    return 300

import re

import re

def extract_runtime_exception_name(stderr_tail: str):
    if not stderr_tail:
        return None
    if "TimeoutExpired" in stderr_tail:
        return "TimeoutExpired"
    m = re.findall(
        r"([A-Za-z_][A-Za-z0-9_]*(?:Error|Exception))\s*:\s*",
        stderr_tail
    )
    return m[-1] if m else None


## 6. Select evaluation samples

Here I randomly sample a fixed number of items from the evaluation split (capped by the eval set size). Since I use a fixed seed and reuse the same `sample_indices`, the **exact same eval subset** is used across models (e.g., Qwen and Mistral), ensuring a fair and consistent comparison.


In [ ]:
# -----------------------
# Choose eval samples
# -----------------------
N_SAMPLES = 88
N_SAMPLES = min(N_SAMPLES, len(eval_dataset))
sample_indices = random.sample(range(len(eval_dataset)), N_SAMPLES)

## 7. Baseline evaluation loop

In this cell, I iterate over the selected eval samples and ask the model to repair each buggy program. I run the model output in an isolated subprocess and record runtime success, error type (or timeout), and LLM latency. If the first attempt fails for a non-timeout reason, I run a second attempt by feeding the runtime error back to the model. Each result is stored in `rows` for later summarization and comparison with other models.


In [ ]:
# -----------------------
# Main evaluation loop (VS Code notebook friendly)
# -----------------------
rows = []

for pos, idx in enumerate(tqdm(sample_indices, desc="Evaluating")):
    ex = eval_dataset[int(idx)]

    exp_exc, exp_line = parse_error_type_field(ex.get("error_type", ""))
    total_llm_time = 0.0

    # ---- Attempt 1
    msgs1 = [
        SystemMessage(content=SYSTEM_TEXT),
        HumanMessage(content=build_user_prompt_no_hint(ex)),
    ]
    pred1, t1 = llm_fix(msgs1)
    total_llm_time += t1
    pred1_clean = extract_python_code(pred1)

    ok1, err1 = run_python(pred1_clean, timeout=timeout_for(ex.get("difficulty", "")))
    err1_type = extract_runtime_exception_name(err1)
    still1 = contains_expected_bug_line(pred1_clean, exp_line)

    # Success on attempt 1
    if ok1:
        rows.append({
            "idx": int(idx),
            "title": ex.get("title", ""),
            "difficulty": ex.get("difficulty", ""),
            "expected_exception": exp_exc,
            "expected_line": exp_line,
            "attempt_used": 1,
            "llm_time_sec": total_llm_time,
            "runtime_ok": True,
            "final_error_tail": "",
            "err1_type": err1_type,
            "err2_type": None,
            "still_bug_line_after_1": still1,
            "still_bug_line_after_2": None,
            "fixed_expected_bug_after_2": True,
            "final_prediction": pred1_clean,
        })
        tqdm.write(f"[OK] {pos+1}/{len(sample_indices)} idx={idx} attempt=1 llm={total_llm_time:.2f}s expected={exp_exc}")
        continue

    # ---- NEW: skip Attempt 2 if Attempt 1 timed out
    if err1_type == "TimeoutExpired":
        rows.append({
            "idx": int(idx),
            "title": ex.get("title", ""),
            "difficulty": ex.get("difficulty", ""),
            "expected_exception": exp_exc,
            "expected_line": exp_line,
            "attempt_used": 1,  # we only tried once
            "llm_time_sec": total_llm_time,
            "runtime_ok": False,
            "final_error_tail": err1,  # store timeout message
            "err1_type": err1_type,
            "err2_type": None,
            "still_bug_line_after_1": still1,
            "still_bug_line_after_2": None,
            "fixed_expected_bug_after_2": None,
            "final_prediction": pred1_clean,
        })
        tqdm.write(
            f"[TIMEOUT] {pos+1}/{len(sample_indices)} idx={idx} attempt=1 llm={total_llm_time:.2f}s "
            f"expected={exp_exc} still_bug_line={still1}"
        )
        continue

    # ---- Attempt 2 (only if attempt 1 failed for a non-timeout reason)
    msgs2 = [
        SystemMessage(content=SYSTEM_TEXT),
        HumanMessage(content=build_user_prompt_with_error(ex, pred1_clean, err1)),
    ]
    pred2, t2 = llm_fix(msgs2)
    total_llm_time += t2
    pred2_clean = extract_python_code(pred2)

    ok2, err2 = run_python(pred2_clean, timeout=timeout_for(ex.get("difficulty", "")))
    err2_type = extract_runtime_exception_name(err2)
    still2 = contains_expected_bug_line(pred2_clean, exp_line)

    # heuristic: did we fix the expected bug?
    if ok2:
        fixed_flag = True
    else:
        fixed_flag = (still2 is False) and (err2_type != exp_exc)

    rows.append({
        "idx": int(idx),
        "title": ex.get("title", ""),
        "difficulty": ex.get("difficulty", ""),
        "expected_exception": exp_exc,
        "expected_line": exp_line,
        "attempt_used": 2,
        "llm_time_sec": total_llm_time,
        "runtime_ok": bool(ok2),
        "final_error_tail": "" if ok2 else err2,
        "err1_type": err1_type,
        "err2_type": err2_type,
        "still_bug_line_after_1": still1,
        "still_bug_line_after_2": still2,
        "fixed_expected_bug_after_2": fixed_flag,
        "final_prediction": pred2_clean,
    })

    status = "OK" if ok2 else "FAIL"
    tqdm.write(
        f"[{status}] {pos+1}/{len(sample_indices)} idx={idx} attempt=2 llm={total_llm_time:.2f}s "
        f"expected={exp_exc} err1={err1_type} err2={err2_type} still_bug_line={still2}"
    )


Evaluating:   1%|          | 1/88 [02:04<3:00:16, 124.33s/it]

[TIMEOUT] 1/88 idx=81 attempt=1 llm=4.13s expected=AttributeError still_bug_line=True


Evaluating:   2%|▏         | 2/88 [05:13<3:52:58, 162.54s/it]

[TIMEOUT] 2/88 idx=14 attempt=1 llm=9.19s expected=NameError still_bug_line=False


Evaluating:   3%|▎         | 3/88 [05:26<2:13:40, 94.36s/it] 

[OK] 3/88 idx=3 attempt=1 llm=7.58s expected=SyntaxError


Evaluating:   5%|▍         | 4/88 [06:21<1:49:57, 78.54s/it]

[OK] 4/88 idx=35 attempt=1 llm=4.71s expected=SyntaxError


Evaluating:   6%|▌         | 5/88 [06:47<1:22:48, 59.86s/it]

[FAIL] 5/88 idx=31 attempt=2 llm=17.99s expected=SyntaxError err1=None err2=None still_bug_line=True


Evaluating:   7%|▋         | 6/88 [06:59<59:18, 43.40s/it]  

[OK] 6/88 idx=28 attempt=1 llm=7.61s expected=IndexError


Evaluating:   8%|▊         | 7/88 [07:29<52:41, 39.03s/it]

[FAIL] 7/88 idx=17 attempt=2 llm=13.14s expected=TypeError err1=None err2=None still_bug_line=True


Evaluating:   9%|▉         | 8/88 [07:51<44:52, 33.66s/it]

[OK] 8/88 idx=13 attempt=1 llm=7.54s expected=NameError


Evaluating:  10%|█         | 9/88 [08:22<43:15, 32.86s/it]

[FAIL] 9/88 idx=69 attempt=2 llm=18.38s expected=SyntaxError err1=None err2=None still_bug_line=True


Evaluating:  11%|█▏        | 10/88 [08:40<36:43, 28.25s/it]

[OK] 10/88 idx=11 attempt=2 llm=11.92s expected=AttributeError err1=None err2=None still_bug_line=False


Evaluating:  12%|█▎        | 11/88 [08:48<28:15, 22.02s/it]

[OK] 11/88 idx=75 attempt=1 llm=4.71s expected=NameError


Evaluating:  14%|█▎        | 12/88 [09:43<40:28, 31.95s/it]

[OK] 12/88 idx=54 attempt=1 llm=8.97s expected=SyntaxError


Evaluating:  15%|█▍        | 13/88 [09:55<32:22, 25.90s/it]

[OK] 13/88 idx=4 attempt=1 llm=8.07s expected=KeyError


Evaluating:  16%|█▌        | 14/88 [10:20<31:38, 25.66s/it]

[FAIL] 14/88 idx=85 attempt=2 llm=13.35s expected=KeyError err1=None err2=None still_bug_line=True


Evaluating:  17%|█▋        | 15/88 [10:32<26:21, 21.67s/it]

[OK] 15/88 idx=78 attempt=1 llm=9.11s expected=SyntaxError


Evaluating:  18%|█▊        | 16/88 [10:44<22:34, 18.82s/it]

[OK] 16/88 idx=27 attempt=1 llm=4.01s expected=ValueError


Evaluating:  19%|█▉        | 17/88 [12:19<49:05, 41.49s/it]

[OK] 17/88 idx=29 attempt=1 llm=9.13s expected=NameError


Evaluating:  20%|██        | 18/88 [12:34<39:11, 33.60s/it]

[OK] 18/88 idx=64 attempt=1 llm=9.75s expected=NameError


Evaluating:  22%|██▏       | 19/88 [12:52<33:21, 29.01s/it]

[FAIL] 19/88 idx=74 attempt=2 llm=18.04s expected=AttributeError err1=None err2=None still_bug_line=False


Evaluating:  23%|██▎       | 20/88 [13:03<26:41, 23.55s/it]

[OK] 20/88 idx=25 attempt=1 llm=5.55s expected=AttributeError


Evaluating:  24%|██▍       | 21/88 [13:31<27:51, 24.95s/it]

[FAIL] 21/88 idx=53 attempt=2 llm=14.25s expected=IndexError err1=None err2=None still_bug_line=False


Evaluating:  25%|██▌       | 22/88 [13:54<26:37, 24.21s/it]

[OK] 22/88 idx=82 attempt=1 llm=4.37s expected=TypeError


Evaluating:  26%|██▌       | 23/88 [15:16<45:04, 41.61s/it]

[FAIL] 23/88 idx=57 attempt=2 llm=18.25s expected=LogicError err1=None err2=None still_bug_line=True


Evaluating:  27%|██▋       | 24/88 [15:24<33:42, 31.60s/it]

[OK] 24/88 idx=84 attempt=1 llm=3.69s expected=LogicError


Evaluating:  28%|██▊       | 25/88 [15:35<26:41, 25.43s/it]

[OK] 25/88 idx=0 attempt=1 llm=7.22s expected=NameError


Evaluating:  30%|██▉       | 26/88 [15:54<24:13, 23.45s/it]

[FAIL] 26/88 idx=48 attempt=2 llm=15.25s expected=SyntaxError err1=None err2=None still_bug_line=True


Evaluating:  31%|███       | 27/88 [16:03<19:32, 19.23s/it]

[OK] 27/88 idx=51 attempt=1 llm=4.27s expected=NameError


Evaluating:  32%|███▏      | 28/88 [16:16<17:14, 17.25s/it]

[FAIL] 28/88 idx=10 attempt=2 llm=12.39s expected=SyntaxError err1=None err2=None still_bug_line=True


Evaluating:  33%|███▎      | 29/88 [18:00<42:26, 43.16s/it]

[OK] 29/88 idx=44 attempt=1 llm=9.34s expected=NameError


Evaluating:  34%|███▍      | 30/88 [18:27<37:03, 38.34s/it]

[FAIL] 30/88 idx=72 attempt=2 llm=12.96s expected=TypeError err1=None err2=None still_bug_line=False


Evaluating:  35%|███▌      | 31/88 [18:46<31:00, 32.65s/it]

[OK] 31/88 idx=21 attempt=1 llm=5.18s expected=TypeError


Evaluating:  36%|███▋      | 32/88 [19:07<27:16, 29.21s/it]

[FAIL] 32/88 idx=87 attempt=2 llm=15.90s expected=KeyError err1=None err2=None still_bug_line=True


Evaluating:  38%|███▊      | 33/88 [19:50<30:30, 33.28s/it]

[FAIL] 33/88 idx=9 attempt=2 llm=15.33s expected=ImportError err1=None err2=None still_bug_line=False


Evaluating:  39%|███▊      | 34/88 [24:39<1:38:55, 109.92s/it]

[OK] 34/88 idx=80 attempt=1 llm=10.61s expected=ImportError


Evaluating:  40%|███▉      | 35/88 [25:59<1:29:17, 101.09s/it]

[OK] 35/88 idx=62 attempt=1 llm=9.95s expected=SyntaxError


Evaluating:  41%|████      | 36/88 [26:21<1:06:53, 77.18s/it] 

[OK] 36/88 idx=65 attempt=1 llm=7.62s expected=NameError


Evaluating:  42%|████▏     | 37/88 [26:30<48:18, 56.84s/it]  

[OK] 37/88 idx=6 attempt=1 llm=3.85s expected=ValueError


Evaluating:  43%|████▎     | 38/88 [26:59<40:28, 48.57s/it]

[FAIL] 38/88 idx=5 attempt=2 llm=11.89s expected=NameError err1=None err2=None still_bug_line=True


Evaluating:  44%|████▍     | 39/88 [27:20<32:48, 40.18s/it]

[FAIL] 39/88 idx=24 attempt=2 llm=14.66s expected=KeyError err1=None err2=None still_bug_line=True


Evaluating:  45%|████▌     | 40/88 [27:31<25:10, 31.46s/it]

[OK] 40/88 idx=61 attempt=1 llm=3.68s expected=ValueError


Evaluating:  47%|████▋     | 41/88 [27:39<19:06, 24.40s/it]

[OK] 41/88 idx=22 attempt=1 llm=4.74s expected=KeyError


Evaluating:  48%|████▊     | 42/88 [28:05<19:05, 24.90s/it]

[FAIL] 42/88 idx=47 attempt=2 llm=18.19s expected=AttributeError err1=None err2=None still_bug_line=False


Evaluating:  49%|████▉     | 43/88 [28:16<15:30, 20.68s/it]

[OK] 43/88 idx=38 attempt=1 llm=4.90s expected=LogicError


Evaluating:  50%|█████     | 44/88 [29:57<32:58, 44.96s/it]

[FAIL] 44/88 idx=16 attempt=2 llm=11.53s expected=KeyError err1=None err2=None still_bug_line=True


Evaluating:  51%|█████     | 45/88 [30:09<25:09, 35.10s/it]

[OK] 45/88 idx=2 attempt=1 llm=4.23s expected=LogicError


Evaluating:  52%|█████▏    | 46/88 [30:41<23:49, 34.04s/it]

[FAIL] 46/88 idx=71 attempt=2 llm=16.17s expected=IndexError err1=None err2=None still_bug_line=False


Evaluating:  53%|█████▎    | 47/88 [31:00<20:11, 29.54s/it]

[FAIL] 47/88 idx=34 attempt=2 llm=11.79s expected=ValueError err1=None err2=None still_bug_line=True


Evaluating:  55%|█████▍    | 48/88 [31:07<15:04, 22.62s/it]

[OK] 48/88 idx=7 attempt=1 llm=3.37s expected=SyntaxError


Evaluating:  56%|█████▌    | 49/88 [31:14<11:45, 18.08s/it]

[OK] 49/88 idx=49 attempt=1 llm=3.52s expected=AttributeError


Evaluating:  57%|█████▋    | 50/88 [31:41<13:08, 20.75s/it]

[FAIL] 50/88 idx=50 attempt=2 llm=14.38s expected=LogicError err1=None err2=None still_bug_line=False


Evaluating:  58%|█████▊    | 51/88 [31:57<11:53, 19.30s/it]

[FAIL] 51/88 idx=70 attempt=2 llm=9.81s expected=AttributeError err1=None err2=None still_bug_line=True


Evaluating:  59%|█████▉    | 52/88 [32:15<11:19, 18.87s/it]

[FAIL] 52/88 idx=18 attempt=2 llm=17.62s expected=SyntaxError err1=None err2=None still_bug_line=True


Evaluating:  60%|██████    | 53/88 [32:24<09:19, 15.98s/it]

[FAIL] 53/88 idx=23 attempt=2 llm=9.03s expected=SyntaxError err1=None err2=None still_bug_line=True


Evaluating:  61%|██████▏   | 54/88 [34:32<28:06, 49.61s/it]

[OK] 54/88 idx=12 attempt=1 llm=5.48s expected=ValueError


Evaluating:  62%|██████▎   | 55/88 [34:54<22:44, 41.35s/it]

[FAIL] 55/88 idx=77 attempt=2 llm=13.55s expected=ValueError err1=None err2=None still_bug_line=True


Evaluating:  64%|██████▎   | 56/88 [35:16<18:53, 35.42s/it]

[FAIL] 56/88 idx=43 attempt=2 llm=15.32s expected=ValueError err1=None err2=None still_bug_line=True


Evaluating:  65%|██████▍   | 57/88 [36:36<25:11, 48.77s/it]

[FAIL] 57/88 idx=86 attempt=2 llm=17.19s expected=ValueError err1=None err2=None still_bug_line=False


Evaluating:  66%|██████▌   | 58/88 [37:03<21:10, 42.35s/it]

[FAIL] 58/88 idx=39 attempt=2 llm=12.98s expected=ValueError err1=None err2=None still_bug_line=False


Evaluating:  67%|██████▋   | 59/88 [38:44<28:59, 59.99s/it]

[OK] 59/88 idx=55 attempt=1 llm=9.13s expected=NameError


Evaluating:  68%|██████▊   | 60/88 [39:23<25:02, 53.65s/it]

[FAIL] 60/88 idx=32 attempt=2 llm=17.21s expected=KeyError err1=None err2=None still_bug_line=True


Evaluating:  69%|██████▉   | 61/88 [39:52<20:49, 46.27s/it]

[FAIL] 61/88 idx=58 attempt=2 llm=16.55s expected=NameError err1=None err2=None still_bug_line=False


Evaluating:  70%|███████   | 62/88 [40:07<16:00, 36.92s/it]

[OK] 62/88 idx=40 attempt=1 llm=8.90s expected=NameError


Evaluating:  72%|███████▏  | 63/88 [40:15<11:46, 28.25s/it]

[OK] 63/88 idx=79 attempt=1 llm=2.97s expected=ValueError


Evaluating:  73%|███████▎  | 64/88 [42:35<24:42, 61.78s/it]

[OK] 64/88 idx=41 attempt=1 llm=4.71s expected=NameError


Evaluating:  74%|███████▍  | 65/88 [42:46<17:49, 46.49s/it]

[OK] 65/88 idx=8 attempt=1 llm=6.14s expected=IndexError


Evaluating:  75%|███████▌  | 66/88 [43:15<15:06, 41.20s/it]

[FAIL] 66/88 idx=83 attempt=2 llm=18.95s expected=NameError err1=None err2=None still_bug_line=True


Evaluating:  76%|███████▌  | 67/88 [43:29<11:35, 33.14s/it]

[OK] 67/88 idx=20 attempt=1 llm=9.01s expected=AttributeError


Evaluating:  77%|███████▋  | 68/88 [43:58<10:37, 31.89s/it]

[FAIL] 68/88 idx=73 attempt=2 llm=14.16s expected=NameError err1=None err2=None still_bug_line=False


Evaluating:  78%|███████▊  | 69/88 [44:17<08:50, 27.93s/it]

[FAIL] 69/88 idx=45 attempt=2 llm=18.37s expected=NameError err1=None err2=None still_bug_line=True


Evaluating:  80%|███████▉  | 70/88 [44:43<08:11, 27.32s/it]

[FAIL] 70/88 idx=52 attempt=2 llm=13.14s expected=AttributeError err1=None err2=None still_bug_line=True


Evaluating:  81%|████████  | 71/88 [44:56<06:30, 23.00s/it]

[OK] 71/88 idx=36 attempt=1 llm=8.33s expected=ImportError


Evaluating:  82%|████████▏ | 72/88 [45:11<05:32, 20.81s/it]

[OK] 72/88 idx=67 attempt=1 llm=8.98s expected=IndexError


Evaluating:  83%|████████▎ | 73/88 [45:20<04:15, 17.06s/it]

[OK] 73/88 idx=37 attempt=1 llm=3.94s expected=IndexError


Evaluating:  84%|████████▍ | 74/88 [45:31<03:34, 15.34s/it]

[OK] 74/88 idx=56 attempt=1 llm=6.84s expected=LogicError


Evaluating:  85%|████████▌ | 75/88 [45:51<03:35, 16.60s/it]

[FAIL] 75/88 idx=60 attempt=2 llm=12.15s expected=ImportError err1=None err2=None still_bug_line=True


Evaluating:  86%|████████▋ | 76/88 [46:41<05:19, 26.63s/it]

[FAIL] 76/88 idx=76 attempt=2 llm=15.57s expected=TypeError err1=None err2=None still_bug_line=False


Evaluating:  88%|████████▊ | 77/88 [51:48<20:18, 110.81s/it]

[FAIL] 77/88 idx=1 attempt=2 llm=7.03s expected=NameError err1=None err2=TimeoutExpired still_bug_line=False


Evaluating:  89%|████████▊ | 78/88 [52:21<14:34, 87.49s/it] 

[FAIL] 78/88 idx=42 attempt=2 llm=19.17s expected=ValueError err1=None err2=None still_bug_line=True


Evaluating:  90%|████████▉ | 79/88 [52:28<09:30, 63.43s/it]

[OK] 79/88 idx=66 attempt=1 llm=3.82s expected=AttributeError


Evaluating:  91%|█████████ | 80/88 [52:38<06:18, 47.32s/it]

[OK] 80/88 idx=15 attempt=1 llm=3.33s expected=AttributeError


Evaluating:  92%|█████████▏| 81/88 [53:09<04:57, 42.47s/it]

[FAIL] 81/88 idx=68 attempt=2 llm=13.74s expected=LogicError err1=None err2=None still_bug_line=False


Evaluating:  93%|█████████▎| 82/88 [53:31<03:37, 36.18s/it]

[FAIL] 82/88 idx=46 attempt=2 llm=18.91s expected=ValueError err1=None err2=None still_bug_line=True


Evaluating:  94%|█████████▍| 83/88 [53:50<02:36, 31.21s/it]

[FAIL] 83/88 idx=26 attempt=2 llm=16.73s expected=NameError err1=None err2=None still_bug_line=True


Evaluating:  95%|█████████▌| 84/88 [53:59<01:38, 24.61s/it]

[OK] 84/88 idx=19 attempt=1 llm=4.90s expected=ValueError


Evaluating:  97%|█████████▋| 85/88 [54:07<00:58, 19.63s/it]

[OK] 85/88 idx=30 attempt=1 llm=3.81s expected=ValueError


Evaluating:  98%|█████████▊| 86/88 [54:25<00:37, 18.86s/it]

[OK] 86/88 idx=33 attempt=1 llm=3.74s expected=AttributeError


Evaluating:  99%|█████████▉| 87/88 [54:34<00:16, 16.06s/it]

[OK] 87/88 idx=63 attempt=1 llm=6.20s expected=NameError


Evaluating: 100%|██████████| 88/88 [54:46<00:00, 37.35s/it]

[OK] 88/88 idx=59 attempt=1 llm=3.35s expected=KeyError


In [15]:
# -----------------------
# Summary + Save (VS Code)
# -----------------------
df = pd.DataFrame(rows)

print("\n=== Summary ===")
print("Total samples:", len(df))
print("Runtime success rate:", round(100 * df["runtime_ok"].mean(), 2), "%")
print("Used attempt=1:", int((df["attempt_used"] == 1).sum()))
print("Used attempt=2:", int((df["attempt_used"] == 2).sum()))
print("Avg LLM time:", round(df["llm_time_sec"].mean(), 2), "sec")


=== Summary ===
Total samples: 88
Runtime success rate: 53.41 %
Used attempt=1: 48
Used attempt=2: 40
Avg LLM time: 10.08 sec


## Conclusion

On the selected evaluation subset (88 samples), the baseline model achieved a **53.41%** runtime success rate. Out of all samples, **48** were fixed on the first attempt, while **40** required a second attempt using runtime error feedback, indicating that error-guided prompting provides a noticeable boost but does not fully resolve all failures. The average model response latency was **10.08 seconds** per sample (including both attempts when used). These results establish the baseline performance that I will compare against the fine-tuned model using the same evaluation split and sample subset.
